### PR TITLE
Vectorized Numpy based functions to Torch based Functions for SpecAugment.

### LINKED ISSUE #10459: [Wav2Vec2] Improve SpecAugment function by converting numpy based function to pytorch based function
# 🚀 Feature request

As can be seen here: https://github.com/huggingface/transformers/blob/11655fafdd42eb56ad94e09ecd84d4dc2d1041ae/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L47,

the function `_compute_mask_indices` (responsible for spec augment) of Wav2Vec2 is written in numpy which means that the function is not GPU compatible. 
The function could simply be rewritten in PyTorch, which should make training on GPU faster.

This "Good First Issue" is about converting `_compute_mask_indices` to PyTorch while keeping the same functionality. 

## Your contribution

I'm happy to guide the contributor through the PR!

### PR DESCRIPTION
**Pull request by 01-vyom**
# What does this PR do?

Fixes #10459 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten

### CODE REVIEW

**Comment by 01-vyom** [comment]
Following are the average test results for the `_compute_mask_indices` function when ran 100 times. Each X.1 subtest case is calculated with `attention_mask = None` and each X.2 subtest case is calculated with `attention_mask` calculated with the following code:
```
attention_mask = torch.ones((batch_size, sequence_length), device=torch_device, dtype=torch.long)
attention_mask[:, -sequence_length // 2 :] = 0
```
1) Test - 100 times
batch_size = 4
sequence_length = 60
mask_prob = 0.5
mask_length = 1
Test **1.1** - 
Result - seconds 
New Code GPU: 0.002872414588928223
New Code CPU: 0.0006633639335632324
Old Code: 0.0003826594352722168
Test **1.2** - 
Result - seconds
New Code GPU: 0.002973439693450928
New Code CPU: 0.0006422805786132813
Old Code: 0.0004153728485107422

2) Test - 100 times
batch_size = 100
sequence_length = 60
mask_prob = 0.5
mask_length = 1
Test **2.1** - 
Result - seconds 
New Code GPU: 0.0663988971710205
New Code CPU: 0.014422652721405029
Old Code: 0.008053600788116455
Test **2.2** - 
Result - seconds
New Code GPU: 0.06568058252334595
New Code CPU: 0.01404146671295166
Old Code: 0.008796172142028809

3) Test - 100 times
batch_size = 1000
sequence_length = 60
mask_prob = 0.5
mask_length = 1
Test **3.1** - 
Result - seconds 
New Code GPU: 0.6623778533935547
New Code CPU: 0.14311392545700075
Old Code: 0.08917582988739013
Test **3.2** - 
Result - seconds
New Code GPU: 0.6566315603256225
New Code CPU: 0.13569485664367675
Old Code: 0.08646429538726806

4) Test - 100 times
batch_size = 4
sequence_length = 1000
mask_prob = 0.5
mask_length = 1
Test **4.1** - 
Result - seconds 
New Code GPU: 0.0031879472732543944
New Code CPU: 0.0013749027252197266
Old Code: 0.00248842716217041
Test **4.2** - 
Result - seconds
New Code GPU: 0.0031322765350341795
New Code CPU: 0.0010571050643920898
Old Code: 0.0015622496604919434

5) Test - 100 times
batch_size = 4
sequence_length = 60
mask_prob = 0.5
mask_length = 4
Test **5.1** - 
Result - seconds 
New Code GPU: 0.003424525260925293
New Code CPU: 0.0008220672607421875
Old Code: 0.0003489851951599121
Test **5.2** - 
Result - seconds
New Code GPU: 0.0034962940216064454
New Code CPU: 0.0007469034194946289
Old Code: 0.0003824186325073242

6) Test - 100 times
batch_size = 4
sequence_length = 1000
mask_prob = 0.5
mask_length = 4
Test **6.1** - 
Result - seconds 
New Code GPU: 0.003502027988433838
New Code CPU: 0.0014672994613647461
Old Code: 0.0017711663246154786
Test **6.2** - 
Result - seconds
New Code GPU: 0.0034971165657043455
New Code CPU: 0.0011277437210083009
Old Code: 0.0011361241340637207

7) Test - 100 times
batch_size = 128
sequence_length = 1000
mask_prob = 0.5
mask_length = 4
Test **7.1** - 
Result - seconds 
New Code GPU: 0.10527128219604492
New Code CPU: 0.04762232780456543
Old Code: 0.052808206081390384
Test **7.2** - 
Result - seconds
New Code GPU: 0.1032623028755188
New Code CPU: 0.03513101100921631
Old Code: 0.03523270606994629

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 1]
```diff
@@ -67,50 +67,43 @@ def _compute_mask_indices(
     Adapted from `fairseq's data_utils.py
     <https://github.com/pytorch/fairseq/blob/e0788f7007a8473a76db573985031f3c94201e79/fairseq/data/data_utils.py#L376>`__.
     """
+    torch_device = "cuda" if torch.cuda.is_available() else "cpu"
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:70]: Actually, people might use "cpu" even when "cuda" is available. Could we instead add a new arg to the function `device` and set it to `device=self.device` (to the model's device) when calling the function?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 2]
```diff
@@ -67,50 +67,43 @@ def _compute_mask_indices(
...
     padding_mask = attention_mask.ne(1) if attention_mask is not None else None
+    if padding_mask is not None:
+        sz = all_sz - padding_mask.long().sum(dim=1)
+        num_mask = (((sz * mask_prob) / mask_length) + torch.rand(bsz, device=torch_device)).int()
+    else:
+        sz = torch.full([bsz], all_sz, device=torch_device)
+        all_num_mask = ((all_sz * mask_prob) / mask_length) + torch.rand(1, device=torch_device)[0]
+        num_mask = torch.full([bsz], all_num_mask, device=torch_device).int()
+
+    num_mask = torch.clamp(num_mask, min=min_masks)
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:83]: any reason to not use `torch.max` here?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 3]
```diff
@@ -67,50 +67,43 @@ def _compute_mask_indices(
...
-        mask_idc = np.random.choice(sz - min_len, num_mask, replace=False)
-        mask_idc = np.asarray([mask_idc[j] + offset for j in range(len(mask_idc)) for offset in range(lengths[j])])
-        mask_idcs.append(np.unique(mask_idc[mask_idc < sz]))
+        curr_num_mask = int(num_mask[i])
+        curr_sz = sz[i]
+        lengths = torch.full([curr_num_mask], mask_length, device=torch_device)
+        if lengths.nelement() == 0:
+            lengths[0] = min(mask_length, curr_sz - 1)
+
+        min_len = lengths[0]
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:91]: why not use `torch.min(...)` here?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 4]
```diff
@@ -67,50 +67,43 @@ def _compute_mask_indices(
...
-        if sz - min_len <= num_mask:
-            min_len = sz - num_mask - 1
-
-        mask_idc = np.random.choice(sz - min_len, num_mask, replace=False)
-        mask_idc = np.asarray([mask_idc[j] + offset for j in range(len(mask_idc)) for offset in range(lengths[j])])
-        mask_idcs.append(np.unique(mask_idc[mask_idc < sz]))
+        curr_num_mask = int(num_mask[i])
+        curr_sz = sz[i]
+        lengths = torch.full([curr_num_mask], mask_length, device=torch_device)
+        if lengths.nelement() == 0:
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:88]: why `.nelement()` instead of of `.sum()`?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 5]
```diff
@@ -67,50 +67,43 @@ def _compute_mask_indices(
...
+        num_mask = (((sz * mask_prob) / mask_length) + torch.rand(bsz, device=torch_device)).int()
+    else:
+        sz = torch.full([bsz], all_sz, device=torch_device)
+        all_num_mask = ((all_sz * mask_prob) / mask_length) + torch.rand(1, device=torch_device)[0]
+        num_mask = torch.full([bsz], all_num_mask, device=torch_device).int()
+
+    num_mask = torch.clamp(num_mask, min=min_masks)
     for i in range(bsz):
-        if padding_mask is not None:
-            sz = all_sz - padding_mask[i].long().sum().item()
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:85]: I think we need to add this code in the loop as well -> `num_mask` depends on the padding_mask which in the new code doesn't seem to be the case

**Review by patrickvonplaten — COMMENTED** [review]
Hey @01-vyom, 

Thanks for tackling this! It's not an easy PR! Can you try to stay as close as possible in terms of operations to the original numpy code? E.g.

`sum(...)` should be translated to `torch.sum(...)` 
`max(...)` should be translated to `torch.max(...)`
....

It would be awesome if you could try to apply the suggestions below. Think then we can do a test run to see that fine-tuning works as expected. 

The speed-up looks very nice!

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 5 above)
Comment (01-vyom) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:85]: In the new code as well as old code the padding mask was calculated beforehand outside the loop.
As padding mask is calculated before the for loop and isn't changing inside the loop we can vectorize that calculation.

In for loop implementation, we calculated the sz and num_mask for each row inside loop.
In the vectorized implementation, we can calculate values of sz and num_mask before for loop.

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 2 above)
Comment (01-vyom) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:83]: As we have vectorized the following process: `num_mask = max(min_masks, num_mask)` checked if a value was greater than a `min_mask` value, if it was we select that number otherwise, we select `min_masks` value, this process was used for a single integer value, but we have a Tensor so to do this operation on all the elements of the tensors, we have performed the [clamp](https://pytorch.org/docs/stable/generated/torch.clamp.html) operation. We could also do something like this: `num_mask[num_mask<min_masks] = min_masks`

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 1 above)
Comment (01-vyom) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:70]: Ok I will create a `device` argument to pass the device type to the function. Can I pass the [`hidden_states.device`](https://github.com/huggingface/transformers/blob/11655fafdd42eb56ad94e09ecd84d4dc2d1041ae/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L842)  as the `device` argument when the `_compute_mask_indices` method is called inside the forward function?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 4 above)
Comment (01-vyom) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:88]: Both the thing are one and the same thing, whenever the lengths variable has zero elements, both will trigger the if condition and the value will be changed. I will change it back to `.sum()` check. But moreover, I will do something like this: `torch.sum(lengths)` as internally `lengths.sum()` calls torch's sum method, and thus will be faster than `lengths.sum()`.

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 3 above)
Comment (01-vyom) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:91]: `torch.min` works only on Tensors,  here we have `mask_length` and `curr_sz-1` as scalar values and thus, `torch.min` will not work.

**Comment by patrickvonplaten** [comment]
Hey @01-vyom,

It looks like the git history is messed up :-/ Sorry about that! This often happens when one does a wrong `git merge` -> could you maybe open a new PR with a clear git history / git diff? Thanks a lot!

**Comment by 01-vyom** [comment]
Ok sure

**Comment by 01-vyom** [comment]
I am closing this PR.
